syntax = "proto3";

package raft;

message VoteRequest {
  int32 term = 1;
  string candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message VoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message LogEntry {
  int32 term = 1;
  int32 index = 2;
  string operation = 3;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
  int32 match_index = 3;
}

message ClientCommand {
  string operation = 1;
}

message ClientReply {
  bool success = 1;
  string leader_id = 2;
  string message = 3;
}

service RaftService {
  rpc RequestVote(VoteRequest) returns (VoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc ClientRequest(ClientCommand) returns (ClientReply);
}
